collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/NumberPickerDialog.java

```diff
@@ -19,76 +19,68 @@
 import androidx.appcompat.app.AlertDialog;
 import android.app.Dialog;
 import android.content.Context;
-import android.content.DialogInterface;
 import android.os.Bundle;
 import androidx.fragment.app.DialogFragment;
+import androidx.lifecycle.ViewModelProvider;
+
 import android.view.LayoutInflater;
 import android.view.View;
 import android.widget.NumberPicker;
 
+import org.jetbrains.annotations.NotNull;
 import org.odk.collect.android.R;
+import org.odk.collect.android.widgets.viewmodels.RangePickerViewModel;
 
 public class NumberPickerDialog extends DialogFragment {
-
-    public static final String NUMBER_PICKER_DIALOG_TAG = "numberPickerDialogTag";
-    public static final String WIDGET_ID = "widgetId";
     public static final String DISPLAYED_VALUES = "displayedValues";
     public static final String PROGRESS = "progress";
 
-    public interface NumberPickerListener {
-        void onNumberPickerValueSelected(int widgetId, int value);
-    }
-
+    private RangePickerViewModel rangePickerViewModel;
     private NumberPickerListener listener;
 
-    public static NumberPickerDialog newInstance(int widgetId, String[] displayedValues, int progress) {
-        Bundle bundle = new Bundle();
-        bundle.putInt(WIDGET_ID, widgetId);
-        bundle.putSerializable(DISPLAYED_VALUES, displayedValues);
-        bundle.putInt(PROGRESS, progress);
-
-        NumberPickerDialog dialogFragment = new NumberPickerDialog();
-        dialogFragment.setArguments(bundle);
-        return dialogFragment;
-    }
-
     @Override
-    public void onAttach(Context context) {
+    public void onAttach(@NotNull Context context) {
         super.onAttach(context);
-
         try {
             listener = (NumberPickerListener) context;
         } catch (ClassCastException e) {
             throw new ClassCastException(context.toString() + " must implement OnHeadlineSelectedListener");
         }
+
+        rangePickerViewModel = new ViewModelProvider(this).get(RangePickerViewModel.class);
+        rangePickerViewModel.setDisplayedValuesForNumberPicker((String[]) getArguments().getSerializable(DISPLAYED_VALUES));
+        rangePickerViewModel.setProgress(getArguments().getInt(PROGRESS));
+
+        rangePickerViewModel.getNumberPickerValue().observe(this, answer -> {
```

Comment:
Why does this value get observed to then just get passed to a listener? Couldn't wherever the listener is trying to get the value (the widget I think) listen? Maybe the data requester should have a method that lets the widget listen for changes and then the data requester can handle the live data subscription (look at `AudioPlayer#onPositionChanged` and its implementation as an example).

Reply:
As another thought that would potentially mean you wouldn't need the `WaitingForDataRegistry` for any of this (as the FormEntryActivity would not have to be involved)